File: Daily/Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y/Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y.cpp
// Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y.cpp
//
// Compile locally with:
// clang++ Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int numberOfSubmatrices(vector<vector<char>> &grid)
    {
        int height = grid.size();
        int width = grid[0].size();
        int count{0};
        vector<vector<pair<int, int>>> dp(height + 1, vector<pair<int, int>>(width + 1, {0, 0}));

        for (int y = 1; y < height + 1; y++)
        {
            for (int x = 1; x < width + 1; x++)
            {
                char cell = grid[y - 1][x - 1];
                dp[y][x].first = dp[y - 1][x].first + dp[y][x - 1].first - dp[y - 1][x - 1].first;
                dp[y][x].second = dp[y - 1][x].second + dp[y][x - 1].second - dp[y - 1][x - 1].second;

                if (cell == 'X')
                {
                    dp[y][x].first++;
                }
                else if (cell == 'Y')
                {
                    dp[y][x].second++;
                }

                if (dp[y][x].first && dp[y][x].first == dp[y][x].second)
                {
                    count++;
                }
            }
        }
        return count;
    }
};

int main()
{
    cout << "Running Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: Daily/Count_Sub_Matricies_with_equal_frequencies_of_X_and_Y/count__sub__matricies_with_equal_frequencies_of__x_and__y.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        height = len(grid)
        width = len(grid[0])
        count = 0

        dp = [[[0, False] for _ in range(width + 1)] for _ in range(height + 1)]
        costMap = {"X": 1, "Y": -1, ".": 0}
        for y in range(1, height + 1):
            for x in range(1, width + 1):
                tmp = grid[y - 1][x - 1]
                dp[y][x][0] = (
                    costMap[tmp]
                    + dp[y - 1][x][0]
                    + dp[y][x - 1][0]
                    - dp[y - 1][x - 1][0]
                )
                if tmp == "X" or dp[y - 1][x][1] or dp[y][x - 1][1]:
                    dp[y][x][1] = True
                if dp[y][x][0] == 0 and dp[y][x][1]:
                    count += 1
        return count


if __name__ == "__main__":
    sol = Solution()
    grid = [[".", "."], [".", "."]]
    print(sol.numberOfSubmatrices(grid))
    print("Running Solution...")
